src/placement/jump_map.c

```diff
@@ -417,11 +417,8 @@ get_target(struct pool_domain *curr_dom, struct pool_target **target,
 			 */
 			range_set = isset_range(dom_used, child_pos, child_pos
 						+ num_doms - 1);
-			if (range_set  && curr_dom->do_children != NULL) {
-				clrbit_range(dom_used, child_pos,
-					     child_pos + (num_doms - 1));
-			}
-
+			if (range_set  && curr_dom->do_children != NULL)
+				break;
```

Comment:
This will clear a single domain if it's used again. 
There may be cases in which there are multiple domain levels  (i.e. racks->servers->nodes->targets) and we expect multiple targets to exist on the same rack but separate servers.

For example if we have 2 Racks, with 8 servers each, with 8 targets each, and we place an object with 8 shards, each Rack will need to have 4 shards, and we will need to clear the the used status for those 2 racks during placement so we can use them again.

Reply:
The current issue with current master is that it will do endless loop here if there no spare targets.  you can try rebuild18 with 4 servers. you can see the issue.

Indeed if we have multiple servers per domain, it will only allow single shard per domain. But for now we only allow single server per domain, see init_pool_metadata.

Reply:
This won't work for widely striped objects though, will it?

If we have a pool map with 4 racks, each with 8 targets, we have a total of 32 targets but the largest object we can place is only 4 shards? Is that correct?

Reply:
yes, only 4 at the moment. otherwise rebuild will not work.

Reply:
Ok that sounds reasonable.

Reply:
> The current issue with current master is that it will do endless loop here if there no spare targets. you can try rebuild18 with 4 servers. you can see the issue.

This is indeed a real problem - and a real bug in the jump map. However, the proposed solution here is not the right one. This will fix this problem, yes, by causing a failure - but will also place dramatic restrictions on how users can configure the system. 

If this is the only problem, we should instead check that all targets are used (i.e. that bitmap is all 1's). If they are used, then abort and return an error because there are no spare targets available.

Reply:
oh, for now, failure domain == single server.  I do not think we support multiple servers in a single domain yet.  If it keeps resetting the domain bits after all has been used, then it will cause co-locate issues, which rebuild can not handle yet. And I suspect there will be other issues as well, need think about it.

Anyway before we support multiple servers for a single domain, I am not sure there is better solution at the moment. I thought about adding another bit-set to track the node, but seems not worth it and also adding un-necessary complexity IMHO.

Reply:
Ok, the problem is definitely communication :)

Let me try to describe more: 

"Domains" can mean many things For example, racks. I know that both the ring map and the jump map support this - the idea of having multiple individual server "domains" within a top-level rack "domain" is definitely supported by the ring map and the jump map and rebuild.

For the purpose of jump map, we consider "co-locating" a shard as meaning placing the same shard on the same target. For a three-level fault domain model (rack, server, target), this means that all three are the same. There is a *separate* bitmap that prevents this from happening. There are two - and this is the wrong one. This bitmap explicitly does not include targets - it includes only the racks and servers. We clear only parts of it at a time, not the whole thing.

For example: we have 2 racks, each with 10 servers containing 1 target each.  We go to place 3 shards. We mark the first rack as used, and the first server as used, and its target as used (target in a different bitmap). Then we use the second rack, and mark it as used. Then the third shard observes that all racks are used, and it clears the "used" marker on the racks, so that a rack can be picked to place on. The reason targets are in a different bitmap is that they are never cleared - so we can not co-locate shards on the same target. 

The fix for the problem you describe above (rebuild not terminating) is to detect when there are no possible targets to place on, and terminate. I suspect it is just re-trying over and over and over, clearing the top-level domain bits but since the bottom one is never cleared, it can never find a location to place on.

Reply:
Oh, sorry, I probably did not make myself clear. I understand what domain means.  And probably both ring and jch support multiple servers under the same domain(rack), but DAOS  does not support that(please check ds_mgmt_pool_create rack == node for the moment), I do not think this would even happen before v1.0.

So at the moment, if jch keeps resetting dom used bits, it will make multiple shards in the same server, which is a bit risky to current DAOS implementation. Yes, for "co-location", I mean put multiple shards in a single server, instead of single target, I understand we will support multiple shards in a single node finally, but we have not test it at all, and I do not think we want to enable it for the moment, at least not for 1.0 IMHO.

Given the constrains above (rack == node, no colocate in single server),  I would suggest to put single shard per domain for the moment to avoid un-expect problems. i.e. not reset the dom set bit.  This is temporary solution before 1.0, of course we can also do a bit nicer, for example add extra bitmap to track node/target used status,  but I am afraid it is not probably not worth for the moment.

Please correct me if I misunderstand sth.

Reply:
Interesting! I didn't realize the support for this wasn't propagated everywhere in DAOS. 

Still though, I think making this change removes functionality that was working more than a year ago. With this change, if you have 8 nodes, and 8 shards/object, and one of those nodes fails, it will become impossible to place that object. This has worked with the ring map for a long time, even with just nodes/targets as domains. 

If `ds_mgmt_pool_create rack == node`, then the placement map should never see "racks" or "nodes" - it should just see nodes. In my example again, for 8 nodes and 8 shards, the code that is currently in the master branch will always guarantee that there are no node-colocations (you will always get 8 different nodes for each layout). But, if one of those 8 nodes fails, you will still be able to recover by placing one of the shards on one of the remaining 7 nodes. I strongly believe this is a good thing - even though it degrades redundancy for a time, it is much better than outright failure. 

We also believe that DAOS-3588 was fixed by 200735f29b8a7a2865405c63f03a52dee647661d, which was merged today.

Reply:
Yes, if there are no spare nodes (not target), it will not be able to rebuild the object (with 8 shards), even if there is spare targets. 

Indeed, for now, pool map can see both racks and nodes, even though racks == nodes, and it is indeed a bit confusing. Yes, I understand jch will always first try to choose different nodes for each shard. But if there are no nodes available, it should either return failure or just put -1 on that shard (sorry forgot the real implementation). Anyway, please do not put multiple shards in the same node for the moment, it is risky and potentially cause problems, since we did not test it at all.   More importantly we do not want to give the user bad impression that DAOS1.0  is not stable, which will hurt us badly IMHO.

Oh, I did not try 200735f yet, but if it still reset the node(rack) bitmap, then it is still possible to put multiple shards in a single node? or I miss sth.  Anyway I will try 200735f to see.

Reply:
As long as there are enough nodes, it will never put multiple shards in a single node. 

It will only place multiple shards on the same node if there is no other choice. I get your point that this will probably result in instability - but the alternative is that the system is completely unusable in this configuration. For example, if you have 8 nodes, and 8 shards, and one fails - the options are: 
1) Fail to find a rebuild target. You are stuck in degraded mode forever - and reintegration will not be possible. The failed node is stuck in DOWN state (because its data can't be rebuilt anywhere) and reintegration can't run (since it runs only on DOWNOUT domains with current design)
2) Select a rebuild target on a different target on one of the remaining nodes. Likely results in instability (as you mentioned before).

I think I finally understand your point now - that being stuck in degraded mode is preferable to  instability. I was thinking in terms of "failure" - but placement will still "work", it'll just return one or more of the slots as -1, but the rest will successfully place. Interesting... I guess I am ok with this if it is really the better of the two options. 

@gnailzenh Can you weigh in on whether we should allow placement to place on multiple targets within the same node? Di suggests that we should not, because it is an untested configuration that may result in 1.0 being unstable for users that attempt to do so.

Reply:
I think node status should be irrelevant to this choice, even if this object is stuck in degraded mode forever, current round of rebuild should be able to finish and mark the node as DOWNOUT.

I prefer to not putting two shards of replicated object into same node (given we support node level redundancy so far), because that will just consume more space (and imbalance)  but won't guarantee anticipated redundancy degree. Though it could boost performance a bit in certain cases (assuming two shards are on different VOS targets of same node), I don't think it's that important comparing with space consuming & anticipated redundancy degree.

The ideal implementation is to make layout algorithm makes right decision (allow co-location in certain domain or not) based on the specified 'redundancy level', that'll be easier for us to support rack level (or VOS target level) replication in the future.

Reply:
> I think node status should be irrelevant to this choice, even if this object is stuck in degraded mode forever, current round of rebuild should be able to finish and mark the node as DOWNOUT. 
Fair enough - my understanding of rebuild was that DOWNOUT = "contents rebuilt". It's certainly possible that we can have stuff move to DOWNOUT even when it is impossible to rebuild its data, though it will be more difficult to implement reintegration. 

> because that will just consume more space but won't guarantee anticipated redundancy degree.
This isn't quite right. The current algorithm (which allows placement multiple times on the same node) **always** provides better redundancy than not allowing this. The reason is that the algorithm will only place on a node more than once if every single node has already been placed on. The alternative (proposed here) is that once every single node has already been placed on, refuse to place any more (fail / return -1). We only place on the same node as a "last resort", providing lower-quality redundancy but still higher than not doing so at all. This is required for widely striped objects - unless we want to define "widely striped" as only one placement on each node, rather than one placement on each available target.

Reply:
> Fair enough - my understanding of rebuild was that DOWNOUT = "contents rebuilt". It's certainly possible that we can have stuff move to DOWNOUT even when it is impossible to rebuild its data, though it will be more difficult to implement reintegration.

I'm not quite sure if it'll make reintegration more complex, but this is what we have to according to current rebuild design. Let's imagine multiple failures case: Rebuild triggered for target A failed at fseq 1, before rebuild finished, target B failed at fseq 2, if the spare target of an object's failed shard is calculated on target B (with higher fseq), then we'll skip this object and delay the object rebuild on next round of rebuild for target B failure (instead of restart current rebuild). See determine_valid_spares()).  

> This isn't quite right. The current algorithm (which allows placement multiple times on the same node) **always** provides better redundancy than not allowing this.

I see your point. My opinion is that the faulty degree (2-way or 3-way replication) should stick to certain fault domain (rack, node or VOS/SSD), it can't be mixed. Your example of 'wide striped object' is actually not replication, and it has no faulty degree requirement, so we can put multiple shards on same node in that case.

Reply:
Took me a bit of digging but I think I understand what's going on. Thanks for pointing me in the right direction, Niu. 

Here's the full process:

- target A fails at fseq 1
- rebuild 1 process starts
- target A would be remapped to target B - note importantly that this requires that target B does not have any other shards on it. Jump map will never remap a shard to a co-located target (existing master branch logic). Unless all other nodes have been used already, it is also guaranteed to be on an unused node. 
- target B fails at fseq 2 while rebuild 1 is still running
- rebuild leader detects the failure and queues another rebuild 2 to run after this rebuild completes
- rebuild 1 tries to place the object at fseq 1, compares against target B which has fseq 2 in determine_valid_spares(), resulting in a -1 (no valid place to rebuild this object this fseq)
- rebuild 1 completes at fseq 1, marking target A as DOWNOUT, but not all of its data redundancy was restored because the spare target for the object also failed during the rebuild. 
- rebuild 2 starts at fseq 2
- rebuild 2 checks placement for the object - now that we're at fseq 2, the fail sequence is not greater than target B, so we loop - and find that the next spare is target C. 
- object ID is sent to target C, which rebuilds/pulls the data.
- rebuild 2 completes at fseq 2, marking target B as DOWNOUT

>  My opinion is that the faulty degree (2-way or 3-way replication) should stick to certain fault domain (rack, node or VOS/SSD), it can't be mixed. 

Can you explain a bit more what you mean here?

> Your example of 'wide striped object' is actually not replication, and it has no faulty degree requirement, so we can put multiple shards on same node in that case.

Fair enough - though right now the jump map currently makes no real differentiation between redundancy shards and striping shards. It places them both the same way - as far "apart" as possible.

Reply:
> Here's the full process:

Yes, exactly. When we found a spare target with higher fseq during rebuild, we don't know if we have already rebuilt something of current object on this newly failed target, so we can only restart current rebuild (this way will introduce lot of dup work and may lead to endless rebuild) or skip this object and wait for next round of rebuild to fix it.

> Can you explain a bit more what you mean here?

I mean if the object is 2-way replication, we have to make sure the two shards of the object are on different node (given we support only node level replication so far)

> Fair enough - though right now the jump map currently makes no real differentiation between redundancy shards and striping shards. It places them both the same way - as far "apart" as possible.

Right, that's something needs be fixed. Our placement algorithm should be aware of the redundancy requirement from the object.